Core/Services/StitchCoreRemoteMongoDBService/Sources/StitchCoreRemoteMongoDBService/Sync/DataSynchronizer.swift

```diff
@@ -482,8 +622,99 @@ public class DataSynchronizer: NetworkStateDelegate, FatalErrorListener {
+        }
+        let lock = nsConfig.nsLock
+        lock.writeLock()
+        defer { lock.unlock() }
+
+        // fetch all of the documents that this filter will match
+        let localCollection = try self.localCollection(for: namespace, withType: Document.self)
+        let beforeDocuments = try localCollection.find(filter)
+
+        // use the matched ids from prior to create a new filter.
+        // this will prevent any race conditions if documents were
+        // inserted between the prior find
+        let ids = beforeDocuments.map({$0["_id"]})
+        var updatedFilter = (options?.upsert ?? false) ? filter :
+            ["_id": ["$in": ids] as Document] as Document
+
+        // do the bulk write
+        let result = try localCollection.updateMany(filter: updatedFilter,
+                                                    update: update,
+                                                    options: options)
+
+        // if this was an upsert, create the post-update filter using
+        // the upserted id.
+        if let upsertedId = result?.upsertedId {
+            updatedFilter = ["_id": upsertedId.value] as Document
+        }
+
+        let upsert = options?.upsert ?? false
+        // iterate over the after-update docs using the updated filter
+        let eventsToEmit: [ChangeEvent<Document>] =
+            try localCollection.find(updatedFilter).compactMap { unsanitizedAfterDocument in
+            // get the id of the after-update document, and fetch the before-update
+            // document from the map we created from our pre-update `find`
+            guard let documentId = unsanitizedAfterDocument["_id"] else {
+                return nil
+            }
+
+            let beforeDocument = beforeDocuments.first(where: {
+                bsonEquals($0["_id"], documentId)
+            })
+
+            // if there was no before-update document and this was not an upsert,
+            // a document that meets the filter criteria must have been
+            // inserted or upserted asynchronously between this find and the update.
+            if beforeDocument == nil && !upsert {
+                return nil
+            }
+
+            // Ensure that the update didn't add any forbidden fields to the document, and remove
+            // them if it did.
+            let afterDocument =
+                try DataSynchronizer.sanitizeCachedDocument(unsanitizedAfterDocument,
+                                                            documentId: documentId,
+                                                            in: localCollection)
+
+            var config: CoreDocumentSynchronization
+            let event: ChangeEvent<Document>
+
+            // if there was no earlier document and this was an upsert,
+            // treat the upsert as an insert, as far as sync is concerned
+            // else treat it as a standard update
+            if let beforeDocument = beforeDocument, !upsert {
```

Comment:
[nit] the `!upsert` check here is irrelevant, since if `beforeDocument != nil`, we know `upsert` is true since the check on line 673 returns if `beforeDocument` is nil and `upsert` is false. I got confused at first because the logic here by itself implies that if upsert is not true and beforeDocument is nil, we still emit an insert event.